tracer/src/Datadog.Trace/SpanExtensions.cs

```diff
@@ -30,12 +36,31 @@ public static void SetUser(this ISpan span, UserDetails userDetails)
             }
 
             var localRootSpan = span;
+            TraceContext traceContext = null;
             if (span is Span spanClass)
             {
-                localRootSpan = spanClass.Context.TraceContext?.RootSpan ?? span;
+                traceContext = spanClass.Context.TraceContext;
+                localRootSpan = traceContext?.RootSpan ?? span;
             }
 
-            localRootSpan.SetTag(Tags.User.Id, userDetails.Id);
+            if (userDetails.PropagateId)
+            {
+                var base64UserId = Convert.ToBase64String(Encoding.UTF8.GetBytes(userDetails.Id));
```

Comment:
Is it worth doing a ballpark check here before we do the allocating. e.g. if userDetails.Id.Length > propagationHeaderMaxLength then we _know_ it won't fit, before we do this. May not be worth it, just a thought

Reply:
done in 7395cb05103cb34e0d4c56a91c41d503e51824a5